contracts/oracle/PCVOracle.sol

```diff
@@ -0,0 +1,370 @@
+    }
+
+    /// @notice check if a venue is in the list of liquid or illiquid venues
+    /// @param venue address to check
+    /// @return boolean whether or not the venue is part of the liquid or illiquid venue list
+    function isVenue(address venue) public view returns (bool) {
+        return liquidVenues.contains(venue) || illiquidVenues.contains(venue);
+    }
+
+    /// @notice get the total PCV balance by looping through the liquid and illiquid pcv deposits
+    /// @dev this function is meant to be used offchain, as it is pretty gas expensive. It also reads
+    /// the fresh balance and not the resistant balance of PCVDeposits, which could be subject to
+    /// in-block manipulations.
+    function getTotalPcv()
+        external
+        view
+        returns (
+            uint256 liquidPcv,
+            uint256 illiquidPcv,
+            uint256 totalPcv
+        )
+    {
+        uint256 liquidVenueLength = liquidVenues.length();
+        uint256 illiquidVenueLength = illiquidVenues.length();
+
+        /// there will never be more than 100 total venues
+        /// so keep the math unchecked to save on gas
+        unchecked {
+            for (uint256 i = 0; i < liquidVenueLength; i++) {
+                address depositAddress = liquidVenues.at(i);
+                (Decimal.D256 memory oracleValue, bool oracleValid) = IOracle(
+                    tokenToOracle[depositToToken[depositAddress]]
+                ).read();
+                require(oracleValid, "PCVO: invalid oracle value");
+
+                liquidPcv +=
+                    (oracleValue.asUint256() *
+                        PCVDeposit(depositAddress).balance()) /
+                    1e18;
+            }
+
+            for (uint256 i = 0; i < illiquidVenueLength; i++) {
+                address depositAddress = illiquidVenues.at(i);
+                (Decimal.D256 memory oracleValue, bool oracleValid) = IOracle(
+                    tokenToOracle[depositToToken[depositAddress]]
+                ).read();
+                require(oracleValid, "PCVO: invalid oracle value");
+
+                illiquidPcv +=
+                    (oracleValue.asUint256() *
+                        PCVDeposit(depositAddress).balance()) /
+                    1e18;
+            }
+
+            totalPcv = liquidPcv + illiquidPcv;
+        }
+    }
+
+    /// ------------- PCV Deposit Only API -------------
+
+    /// @notice update the cumulative and last updated times
+    /// only callable by a liquid pcv deposit
+    /// this allows for lazy evaluation of the TWAPCV
+    /// @param pcvDelta the amount of PCV change in the venue
+    function updateLiquidBalance(int256 pcvDelta)
+        external
+        onlyVoltRole(VoltRoles.LIQUID_PCV_DEPOSIT_ROLE)
+    {
+        address token = PCVDeposit(msg.sender).balanceReportedIn();
+        IOracle oracle = IOracle(tokenToOracle[token]);
+        (Decimal.D256 memory oracleValue, bool oracleValid) = oracle.read();
```

Comment:
This feels like a lot of steps that we can consolidate into this contract into a mapping of an address of a pcv deposit to a struct. This would allow us to save 2 CALLS + 2 SLOADS which is at least 5 thousand gas per swap and instead make it only a single SLOAD to get the decimal normalizer. I don't think we're going to be having volatile assets in the PCV any time soon.

Reply:
not necessarily volatile assets, but at some point the protocol could have LUSD, VOLT-USDC AMM LP tokens, cTokens, etc, and it would be more correct for instance for a Compound PCVDeposit to report a balance in cToken balance & then have an oracle return a price of 1.05$

Reply:
That makes sense. The PCV deposits do have this information on pricing, however that seems clunky from an architecture perspective to have the PCV deposit pass the price to the PCV Oracle. This current design makes more sense because it separates concerns properly.